src/encord_active/app/project_comparison/project_similarity.py

```diff
@@ -0,0 +1,165 @@
+    metrics_filtered = []
+    for metric_name in all_metrics:
+        if metric_name in merged_metrics_1 and merged_metrics_1[metric_name].dropna().shape[0] > 0:
+            metrics_filtered.append(metric_name)
+
+    if len(metrics_filtered) <= 1:
+        st.write("At least two metrics should be generated for 2D metric similarity plot")
+        return
+
+    metric_selection_col_1, metric_selection_col_2 = st.columns(2)
+    metric_name_1 = metric_selection_col_1.selectbox(
+        "Select the first metric",
+        options=metrics_filtered,
+        index=0,
+        key=f"project_comparison_metric_selection_1_{get_state().project_paths.project_dir.name}_{project_name_2}",
+    )
+
+    # Second selectbox should be populated according to the selected metric in the first one
+    # due to different metric types. E.g., 'Area' is not compatible with 'object aspect ratio'
+    options_for_second_metric = []
+
+    for metric_name in metrics_filtered:
+        if metric_name == metric_name_1:
+            continue
+        metric_df_1 = merged_metrics_1[metric_name_1].dropna().to_frame()
+        metric_df_2 = merged_metrics_1[metric_name].dropna().to_frame()
+
+        merged_metric = metric_df_1.join(metric_df_2, how="inner")
+        if merged_metric.shape[0] > 0:
+            options_for_second_metric.append(metric_name)
+
+    if len(options_for_second_metric) == 0:
+        st.write("There is no compatible metric")
+        return
+
+    metric_name_2 = metric_selection_col_2.selectbox(
+        "Select the second metric",
+        options=options_for_second_metric,
+        key=f"project_comparison_metric_selection_2_{get_state().project_paths.project_dir.name}_{project_name_2}",
+    )
+
+    if metric_name_1 == metric_name_2:
+        st.warning("Please choose a different metric")
+        return
+
+    project_values_1 = merged_metrics_1[[metric_name_1, metric_name_2]].copy().dropna()
+    project_name_1 = get_state().project_paths.project_dir.name
+    project_values_1["project"] = project_name_1
+
+    # Check if the selected metrics are also available in the second project
+    if metric_name_1 not in merged_metrics_2 or merged_metrics_2[metric_name_1].dropna().shape[0] == 0:
+        st.write(f"Metric **{metric_name_1}** is not available for the project `{project_name_2}`.")
+        return
+
+    if metric_name_2 not in merged_metrics_2 or merged_metrics_2[metric_name_2].dropna().shape[0] == 0:
+        st.write(f"Metric **{metric_name_2}** is not available for the project `{project_name_2}`.")
+        return
+
+    project_values_2 = merged_metrics_2[[metric_name_1, metric_name_2]].copy().dropna()
+    project_values_2["project"] = project_name_2
+
+    # TODO change append to concat
+    project_values = pd.concat([project_values_1, project_values_2], ignore_index=True)
+
+    fig = render_2d_metric_similarity_plot(
+        project_values, cast(str, metric_name_1), cast(str, metric_name_2), project_name_1, project_name_2
```

Comment:
If we type the metric lists above, we won't have to cast anything here.
Generally, when we start casting something, it should raise an internal alarm. It shouldn't be necessary.

Reply:
You are certainly right about casting, however what happens in this case is that when streamlit returns the selected name as below:

```
    metric_name_1 = metric_selection_col_1.selectbox(
        "Select the first metric",
        options=metrics_filtered,
        index=metrics_filtered.index("Brightness") if "Brightness" in metrics_filtered else 0,
        key=f"project_comparison_metric_selection_1_{get_state().project_paths.project_dir.name}_{project_name_2}",
    )
```

the `metric_name_1` is of type `Optional[str]`; therefore mypy raises an error in this line, that is why I had to restrict its type. Another solution might be checking the type as below just before line 103?

```
instance(metric_name_1, str)
instance(metric_name_2, str)
```

What do you think?

Reply:
Checking for none is what we should do. It's the safest.